### PR TITLE
Feature/implement identity microservice

### PR DESCRIPTION
**Pull request by denis-pptx**

### CODE REVIEW

**Comment by denis-pptx** [comment]
Points to be corrected:
1) Use Global exception handler instead of a lot
2) Add FromQuery / FromBody / ... to Actions in Controllers 
3) Add tokens expiry time to appsettings.json
4) Manual mapping or specifying properties for mapping
5) Add messages to custom exceptions

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserVm.cs [hunk 1]
```diff
@@ -0,0 +1,3 @@
+﻿namespace Identity.Application.Features.Identity.Commands.LoginUser;
+
+public record LoginUserVm(string AccessToken, string RefreshToken);
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserVm.cs]: let's make this entity a dto object and move it to the dtos folder

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserCommand.cs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+﻿namespace Identity.Application.Features.Identity.Commands.LoginUser;
+
+public record LoginUserCommand(string Email, string Password)
+    : ICommand<LoginUserVm>;
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserCommand.cs]: the dto must be passed to the command

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserHandler.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
+﻿namespace Identity.Application.Features.Identity.Commands.LoginUser;
+
+public class LoginUserHandler(UserManager<User> _userManager, IJwtProvider _jwtProvider)
+    : ICommandHandler<LoginUserCommand, LoginUserVm>
+{
+    public async Task<LoginUserVm> Handle(LoginUserCommand request, CancellationToken cancellationToken)
+    {
+        var user = await _userManager.FindByEmailAsync(request.Email);
+        if (user is null)
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserHandler.cs]: there should be an empty line here

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserHandler.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            throw new UnauthorizedException("Incorrect Email or Password");
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserHandler.cs]: move constants to a separate file

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserHandler.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        bool isPasswordValid = await _userManager.CheckPasswordAsync(user, request.Password);
+        if (!isPasswordValid)
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserHandler.cs]: there should be an empty line here

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserHandler.cs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            throw new UnauthorizedException("Incorrect Email or Password");
+        }
+
+        var accessToken = await _jwtProvider.GenerateAccessTokenAsync(user);
+        var refreshTokenModel = _jwtProvider.GenerateRefreshToken();
+
+        user.RefreshToken = refreshTokenModel.Token;
+        user.RefreshExpiryTime = refreshTokenModel.ExpiryTime;
+        await _userManager.UpdateAsync(user);
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/LoginUser/LoginUserHandler.cs]: there should be an empty line here

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/RefreshToken/RefreshTokenVm.cs [hunk 7]
```diff
@@ -0,0 +1,3 @@
+﻿namespace Identity.Application.Features.AuthenticaIdentitytion.Commands.RefreshToken;
+
+public record RefreshTokenVm(string AccessToken, string RefreshToken);
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/RefreshToken/RefreshTokenVm.cs]: create dto

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/RefreshToken/RefreshTokenCommand.cs [hunk 8]
```diff
@@ -0,0 +1,4 @@
+﻿namespace Identity.Application.Features.Identity.Commands.RefreshToken;
+
+public record RefreshTokenCommand(string AccessToken, string RefreshToken) 
+    : ICommand<RefreshTokenVm>;
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/RefreshToken/RefreshTokenCommand.cs]: pass dto

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/RefreshToken/RefreshTokenHandler.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+public class RefreshTokenHandler(IJwtProvider _jwtProvider, UserManager<User> _userManager)
+    : ICommandHandler<RefreshTokenCommand, RefreshTokenVm>
+{
+    public async Task<RefreshTokenVm> Handle(RefreshTokenCommand request, CancellationToken cancellationToken)
+    {
+        var principal = _jwtProvider.GetPrincipalFromAccessToken(request.AccessToken);
+
+        var userId = principal?.FindFirst(ClaimTypes.NameIdentifier)?.Value;
+        if (userId is null)
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/RefreshToken/RefreshTokenHandler.cs]: there should be an empty line here

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/RefreshToken/RefreshTokenHandler.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+        {
+            throw new UnauthorizedException();
+        }
+
+        var user = await _userManager.FindByIdAsync(userId);
+        if (user is null ||
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/RefreshToken/RefreshTokenHandler.cs]: tha same

File: src/Services/Identity/Identity.Application/Features/Identity/Commands/RegisterUser/RegisterUserCommand.cs [hunk 11]
```diff
@@ -0,0 +1,4 @@
+﻿namespace Identity.Application.Features.Identity.Commands.RegisterUser;
+
+public record RegisterUserCommand(string UserName, string Email, string Password)
+    : ICommand<Unit>;
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Identity/Commands/RegisterUser/RegisterUserCommand.cs]: pass dto

File: src/Services/Identity/Identity.Application/Features/Users/Commands/DeleteUser/DeleteUserByIdHandler.cs [hunk 12]
```diff
@@ -0,0 +1,24 @@
+﻿namespace Identity.Application.Features.Users.Commands.DeleteUser;
+
+public class DeleteUserByIdHandler(UserManager<User> _userManager, IUserService _userService) 
+    : ICommandHandler<DeleteUserByIdCommand, Unit>
+{
+    public async Task<Unit> Handle(DeleteUserByIdCommand request, CancellationToken cancellationToken)
+    {
+        var userIdentity = await _userManager.FindByIdAsync(request.Id.ToString());
+        if (userIdentity == null)
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Users/Commands/DeleteUser/DeleteUserByIdHandler.cs]: there should be an empty line here

File: src/Services/Identity/Identity.Application/Features/Users/Commands/UpdateUserRole/UpdateUserRoleHandler.cs [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+        }
+
+        var user = await _userManager.FindByIdAsync(request.UserId.ToString());
+        if (user is null)
+        {
+            throw new NotFoundException("User not found");
+        }
+
+        string? currentRole = (await _userManager.GetRolesAsync(user)).SingleOrDefault();
+        string? newRole = (await _roleManager.FindByNameAsync(request.NewRole))?.Name;
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Users/Commands/UpdateUserRole/UpdateUserRoleHandler.cs]: use 'var'

File: src/Services/Identity/Identity.Application/Features/Users/Queries/GetAllUsers/GetAllUsersHandler.cs [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        var users = await _userManager.Users.ToListAsync(cancellationToken);
+        var userVms = new List<UserVm>();
+
+        foreach (var user in users)
+        {
+            string? role = (await _userManager.GetRolesAsync(user)).SingleOrDefault();
+
+            var userVm = _mapper.Map<UserVm>(user);
+            userVm.Role = role ?? string.Empty;
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Users/Queries/GetAllUsers/GetAllUsersHandler.cs]: why might a user not have a role?

File: src/Services/Identity/Identity.Application/Features/Users/Queries/Models/UserVm.cs [hunk 15]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Identity.Application.Features.Users.Queries.Models;
+
+public class UserVm
+{
+    public string Id { get; set; } = string.Empty;
+    public string UserName {  get; set; } = string.Empty;
+    public string Email { get; set; } = string.Empty;
+    public string Role {  get; set; } = string.Empty;   
+}
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Application/Features/Users/Queries/Models/UserVm.cs]: create dto

File: src/Services/Identity/Identity.Infrastructure/Authentication/JwtProvider.cs [hunk 16]
```diff
@@ -0,0 +1,64 @@
+﻿using Identity.Application.Models;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.IdentityModel.JsonWebTokens;
+
+namespace Identity.Infrastructure.Authentication;
+
+public class JwtProvider(
+    IOptions<JwtOptions> _jwtOptions,
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Infrastructure/Authentication/JwtProvider.cs]: move it to Application

File: src/Services/Identity/Identity.Infrastructure/Authentication/JwtOptions.cs [hunk 17]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Identity.Infrastructure.Authentication;
+
+public class JwtOptions
+{
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Infrastructure/Authentication/JwtOptions.cs]: move it to Application

File: src/Services/Identity/Identity.Infrastructure/Data/ApplicationDbContext.cs [hunk 18]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Identity.Infrastructure.Data;
+
+public class ApplicationDbContext : IdentityDbContext<User>
+{
+    public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options)
+            : base(options)
+    {
+        Database.EnsureCreated();
+    }
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.Infrastructure/Data/ApplicationDbContext.cs]: use migration

File: src/Services/Identity/Identity.WebUI/Controllers/AuthController.cs [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+
+[Route("api/[controller]")]
+[ApiController]
+public class AuthController(IMediator _mediator) 
+    : ControllerBase
+{
+    // POST api/<AuthController>/register
+    [HttpPost("register")]
+    public async Task<IActionResult> Register([FromBody] RegisterUserCommand command, CancellationToken cancellationToken)
+    {
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.WebUI/Controllers/AuthController.cs]: dto must be passed as an input parameter

File: src/Services/Identity/Identity.WebUI/Program.cs [hunk 20]
```diff
@@ -1,14 +1,48 @@
+using Identity.Application.Mappings;
+using Identity.Infrastructure.Data;
+using Identity.Infrastructure.Data.Seed;
+using Identity.WebUI.Extensions;
+using Microsoft.EntityFrameworkCore;
+using System.Reflection;
+
 var builder = WebApplication.CreateBuilder(args);
 
 // Add services to the container.
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.WebUI/Program.cs]: extra lines

File: src/Services/Identity/Identity.WebUI/Program.cs [hunk 21]
```diff
@@ -1,14 +1,48 @@
...
 
 builder.Services.AddControllers();
-// Learn more about configuring Swagger/OpenAPI at https://aka.ms/aspnetcore/swashbuckle
+
 builder.Services.AddEndpointsApiExplorer();
 builder.Services.AddSwaggerGen();
 
+//var connection = builder.Configuration.GetConnectionString("MySQL");
+//builder.Services.AddDbContext<ApplicationDbContext>(
+//    options => options.UseMySql(connection, new MySqlServerVersion(new Version(8, 3, 0))));
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.WebUI/Program.cs]: delete comments

File: src/Services/Identity/Identity.WebUI/Program.cs [hunk 22]
```diff
@@ -1,14 +1,48 @@
...
+
+builder.Services.AddMediatR(cfg => cfg.RegisterServicesFromAssemblyContaining(typeof(LoginUserHandler)));
+
+builder.Services.AddExcepitonHandlers();
+
+builder.Services.AddProblemDetails();
+
+builder.Services.AddHttpContextAccessor();
+
+
```
Comment (Elizaveta564) [inline src/Services/Identity/Identity.WebUI/Program.cs]: extra line

**Review by Elizaveta564 — CHANGES_REQUESTED** [review]

File: src/Services/Identity/Identity.Application/Features/Users/Commands/DeleteUser/DeleteUserByIdHandler.cs [hunk 23]
```diff
@@ -0,0 +1,26 @@
+﻿namespace Identity.Application.Features.Users.Commands.DeleteUser;
+
+public class DeleteUserByIdHandler(UserManager<User> _userManager, ICurrentUserService _userService) 
+    : ICommandHandler<DeleteUserByIdCommand, Unit>
+{
+    public async Task<Unit> Handle(DeleteUserByIdCommand request, CancellationToken cancellationToken)
+    {
+        var userIdentity = await _userManager.FindByIdAsync(request.Id.ToString());
+
+        if (userIdentity == null)
```
Comment (mebry) [inline src/Services/Identity/Identity.Application/Features/Users/Commands/DeleteUser/DeleteUserByIdHandler.cs]: is null

File: src/Services/Identity/Identity.Application/Features/Users/Queries/GetUserById/GetUserByIdHandler.cs [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        var user = await _userManager.Users
+            .FirstOrDefaultAsync(x => x.Id == request.Id.ToString(), cancellationToken);
+        
+        if (user is null)
+        {
+            throw new NotFoundException(UserErrorMessages.NotFound);
+        }
+
+        string? role = (await _userManager.GetRolesAsync(user)).Single();
```
Comment (mebry) [inline src/Services/Identity/Identity.Application/Features/Users/Queries/GetUserById/GetUserByIdHandler.cs]: use var

**Review by mebry — APPROVED** [review]

**Pull request merged**